src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs

```diff
@@ -41,6 +41,7 @@ public class CryptoProviderFactory
     {
         private static CryptoProviderFactory _default;
         private static ConcurrentDictionary<string, string> _typeToAlgorithmMap = new ConcurrentDictionary<string, string>();
+        private static object _cacheLock = new object();
```

Comment:
Another approach would be to move the lock into the CryptoProviderCache class, and expose a new method to provide this capability. ConcurrentDictionary demonstrates a good pattern for this scenario via the [GetOrAdd()](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netframework-4.5#System_Collections_Concurrent_ConcurrentDictionary_2_GetOrAdd__0_System_Func__0__1__) function. It either returns a known value or uses the value specified by the caller instead. It includes two overloads, one with an explicit object reference and another that takes a Func<> for the cases where the object is expensive to create.

Reply:
@CBaud what you point out is a better model because we are locking from the outside of the cache. As soon as something other than SignatureProviders are cached OR the cache is shared between types, we will see a problem.

Not sure if we should address this in 5.2.5.
For sure as soon as we cache other crypto operators, such as KeyWrapProvider, AuthenticatedEncryptionProviders, we would smart to change this design.

Reply:
I just added an issue for moving the lock into the CryptoProviderCache: #1004
It's currently set for the 5.2.6 release.